[lucas-sc0] iP

lucas-sc0 opened the pull request.

Commit:
Add increment level 0

Renamed chatbot duke to monday and added greet and exit messages.

Commit:
Revert "Add increment level 0"

This reverts commit a28afc420c7c326365bb63aa49cc8fd6c483b12b.

Commit:
Reapply "Add increment level 0"

This reverts commit 8162d95ecf55bb8e0317edeebbbe48c3fe98f084.

Commit:
Add increment level 1

Improved the skeletal version of Monday so that it echos commands entered by the user, and exits when the user types the command bye.

Commit:
Add increment level 2

Added the ability to store whatever text entered by the user and display them back to the user when requested. Also added indents for Monday's inputs for clarity.

Commit:
Add increment level 3

Added the ability to mark tasks as done.and the ability to change the status back to not done. Also added Task class to handle operation on tasks.

Commit:
Add increment A-CodingStandard

Ensured code complies with standard provided and refactored the code to provide more abstaction and align with OOP.

Commit:
Add increment level 4

Add support for tracking ToDos, Deadlines and Events. Also used Inheritance and polymorphism to support multiple task types.

Commit:
Add increment automated text ui testing

Used the input/output redirection technique to semi-automate the testing of Duke.

Commit:
Add increment a-codequality

Improvedcode quality where necessary. Added constants like MAX_TASKS so there are no magic numbers.

rahuljai-05 commented on src/main/java/Monday.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private void markTaskAsDone(String input) {
+        int taskNumber = Integer.parseInt(input.substring(5)) - 1;
```
Maybe you could add a comment to explain why you parse the input from the 5th index ?
You could probably also just extract the magic number 5 as a constant with a descriptive name to explain what it does eg. INDEX_OF_NUMBER_AFTER_MARK

rahuljai-05 commented on src/main/java/Monday.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private void unmarkTaskAsNotDone(String input) {
+        int taskNumber = Integer.parseInt(input.substring(7)) - 1;
```
Same suggestion as before regarding the parsing index.

rahuljai-05 commented on src/main/java/Monday.java:
```diff
@@ -0,0 +1,98 @@
+            input = scanner.nextLine();
+            System.out.println(LINE);
+
+            if (input.equalsIgnoreCase("list")) {
+                listTasks();
+            } else if (input.startsWith("mark ")) {
+                markTaskAsDone(input);
+            } else if (input.startsWith("unmark ")) {
+                unmarkTaskAsNotDone(input);
+            } else if (input.equalsIgnoreCase("bye")) {
+                printGoodbyeMessage();
+                break;
+            } else {
+                addTask(input);
+            }
```
I like the clear separation of concerns using different methods, all with descriptive names !

rahuljai-05 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+        this.isDone = false;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " ");
+    }
```
I like how this method is named appropriately and also the efficient one line return statement !

rahuljai-05 reviewed (commented):
The coding standard seems to follow all the conventions and it is overall a very good submission.

Commit:
Add MondayException class

Commit:
Add enhanced error handling and task creation validation

- Implemented custom exception handling with `MondayException` to provide specific feedback for invalid task input formats.

- Modified the `addTask` method to throw `MondayException` for cases where the description of a todo, deadline, or event is missing, or if the input format is incorrect.

- Updated `handleUserInput` method to catch and display user-friendly error messages based on different invalid input scenarios.

- Ensured clear instructions are given for invalid commands or empty task descriptions, improving user experience.

Commit:
Add error handling for marking and unmarking tasks

- Enhanced error handling for invalid task numbers and non-integer inputs using try-catch blocks.

- Introduced MondayException to handle custom errors for invalid task numbers.

- Provided user-friendly error messages when task numbers are invalid or improperly formatted.

Commit:
Update test input and expected text

Commit:
Merge branch 'branch-Level-5'

Commit:
Organize the classes into suitable java packages

Commit:
Add new delete method in Monday

Removed Duplicate Declaration of tasks: Only used ArrayList<Task> now.

Implemented Task Deletion: Added deleteTask method and command handling.

Updated Task Handling Methods: Modified markTaskAsDone, unmarkTaskAsNotDone, listTasks, and addTask methods to work with ArrayList<Task>.

Updated Command Handling: Adjusted handleUserInput to include the delete command.

Commit:
Add save and load Task methods

Commit:
Add save tasks and load tasks function

Commit:
Add getter methods for deadline event and task classes

Commit:
Fix bug for set all tasks to mark on load

Commit:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Monday.java

Commit:
Create jar manifest

tzerbin commented on src/main/java/Monday.java:
```diff
@@ -0,0 +1,264 @@
+import java.util.ArrayList;
+
+public class Monday {
+    private static final String LOGO = " __  __                 _             \n"
```
Nice usage of constants!

tzerbin commented on src/main/java/Monday.java:
```diff
@@ -0,0 +1,264 @@
+                    printWriter.println("D | " + (deadline.isDone() ? "1" : "0") + " | " + deadline.getDescription() + " | " + deadline.getBy());
+                } else if (task instanceof Event) {
+                    Event event = (Event) task;
+                    printWriter.println("E | " + (event.isDone() ? "1" : "0") + " | " + event.getDescription() + " | " + event.getFrom() + " | " + event.getTo());
```
Perhaps break this line of code into multiple lines?

tzerbin commented on src/main/java/exception/MondayException.java:
```diff
@@ -0,0 +1,7 @@
+package exception;
+
+public class MondayException extends Exception {
```
What does it mean by "MondayException"? Perhaps you could use a more descriptive name for the exception class.

tzerbin commented on src/main/java/model/Todo.java:
```diff
@@ -0,0 +1,12 @@
+package model;
```
Why ask Task classes packaged in "model"? What does "model" mean here?

tzerbin reviewed (commented):
Very good attempt in following the guidelines!

Only small nits here and there.

tzerbin reviewed (commented):
Very good attempt in following the guidelines!

Only small nits here and there.

Commit:
Use more OOP

Refactored the code to extract out closely related code as classes
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list e.g., it has operations to add/delete tasks in the list
Implemented command classes:
  - `AddCommand` for adding tasks to the list.
  - `DeleteCommand` for removing tasks from the list.
  - `ExitCommand` for terminating the application.
  - `ListCommand` for displaying all tasks.
  - `MarkCommand` for marking tasks as done.
  - `UnmarkCommand` for unmarking completed tasks.

Commit:
Add date conversion to Deadline task

Commit:
Refine datetime conversion function

Commit:
Add find command

lucas-sc0 committed:
Merge pull request #2 from lucas-sc0/branch-Level-9

Add find command

Commit:
Add JavaDoc comments to the code

Commit:
Add a User Guide

Commit:
Update user guide

Commit:
Improved readability of code

Commit:
Update user guide

Commit:
update user guide

Commit:
Update user guide

Commit:
Revert "update user guide"

This reverts commit 9e7765ccdcb7917676c3c9cf266053dee88575e8.

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Revert "update user guide"

This reverts commit 9e7765ccdcb7917676c3c9cf266053dee88575e8.

Commit:
Update user guide

Commit:
Refactor generic MondayException into more descriptive names

Commit:
Refactor model package name into tasktypes

Commit:
Test and finalize Monday-v0.2

Commit:
Remove magic numbers

Commit:
Update url for jar file link
